### ISSUE TITLE
XML files for external loads are not readable

### ISSUE THREAD

**Issue by aymanhab** [opensim-org/opensim-core#2324]
PR https://github.com/opensim-org/opensim-core/pull/2314 changes the layout of ExternalLoads.xml file causing old files not to be readable. This can be checked by creating a new file and diffing against older file.  Must fix for 4.0 since it affects all our and user workflows that use external loads.

**Comment by aymanhab** [comment]
In case it helps, the latest build from master has the same issue (without the rollback of PR #2314) trying to edit an old ExternalLoads.xml file.

**Comment by aseth1** [comment]
I am using the build OpenSim 4.0-2018-10-17-a1ed8256 from [here ](https://ci.appveyor.com/project/opensim-org/opensim-gui/builds/19587105/artifacts) that we tested during the sprint review.

When running InverseDynamics from the GUI, I do see the error in trying to edit (clicking the pencil) and external loads XML file:
![image](https://user-images.githubusercontent.com/6494122/47325453-acb67d00-d618-11e8-9fd8-420e6d72dd53.png)
However, the Tool runs and produces identical results with 3.3.

I then updated the file using *Tools->Convert Files...* and I still get the same error. Therefore, I don't believe it is an updateFromXML issue but some particular expectation by the GUI about the file. Here is the diff:
![image](https://user-images.githubusercontent.com/6494122/47325577-60b80800-d619-11e8-907b-783928e04e95.png)

Again the Tool runs and produces the same results.

I was able to edit the *subject01_walk1_grf.xml* in an earlier build OpenSim_128d4fbd-2018-08-06 and save it as a new ExternalLoads file. That file was editable in the latest build. The diff between these files is:
![image](https://user-images.githubusercontent.com/6494122/47325759-416daa80-d61a-11e8-831e-cd4b98e4c5f9.png)

Consequently, the source of the problem is the path to the files. If we used the absolute path there doesn't seem to be an issue editing it in the GUI. The irony is that the Tool runs and can find the file locally. I do not believe this is an `updateFromXML` issue.

**Comment by aseth1** [comment]
I was also able to make an `ExternalLoads` from scratch without issues and produce the same results from testID case with the *subject01.osim* model.

**Comment by aseth1** [comment]
Simply updating the external loads file in OpenSim_128d4fbd-2018-08-06 (using *Tools->Convert Files...*) does not update the path to absolute. It is the Tool/GUI that is updating the path in OpenSim_128d4fbd-2018-08-06, which does not appear to be happening in OpenSim 4.0-2018-10-17-a1ed8256.

**Comment by aymanhab** [comment]
Thanks for the update @aseth1 


Here's my current findings:
- In the past (pre refactoring of ExternalLoads) invokeConnectToModel was only exercised from AbstractTool::createExternalLoads which is executed in the directory of the xml file so relative paths for data flies worked correctly as they do now on run. 
- Now, however, connection is executed earlier in initSystem after adding the ExternalLoads to a model so it has the wrong context and can't find the data files.

**Comment by chrisdembia** [comment]
I advocate for using `SimTK::Pathname` to load the data source. I can make the PR if necessary.

**Comment by aseth1** [comment]
> Now, however, connection is executed earlier in initSystem after adding the ExternalLoads to a model so it has the wrong context and can't find the data files.

The Tool executes fine and produces the correct results so it is not that the data files cannot be found, because the `ExternalLoads` is finding and loading the data without issues. How does the GUI locate the source file when the user selects to edit it? There seems to be a mismatch between the two.

**Comment by aseth1** [comment]
Thanks @chrisdembia  if that is in fact the issue then it seems like the proper fix.

**Comment by aymanhab** [comment]
Let me explain @aseth1 
1. Before 4.0 ExternalLoads was not a "Component" so we could create it from file, pass Model to the constructor and be done (edit ExternalForcess etc.)
2. In 4.0 ExternalLoads became a component so we were required to create it and add it to a model in some fashion and call initSystem on the model so we could use it. We could get away with that before the last change because invokeConnectToModel wasn't triggered then. This pathway was used by the GUI to load/edit/define ExternalLoads.
3. When refactoring ExternalLoads to be "Proper component", invokeConnectToModel was renamed and is now actually called through initSystem but that is done in context/working-directory different from  the call in createExternalLoads (which changes working directory around the call) resulting in different behavior in GUI edit vs tool run.

We need to make sure the two instances of "ExternalLoads::extendConnect...." during construction of ExternalLoads and the call from AbstractTool::createExternalLoads behave identically regardless of working directory which is not the case now if the .mot file is specified using relative path.

The above explains every single issue you and I experienced, and will not be changed by moving to any other tool/utility, unless we forbid using relative path for .mot file specification. Please let me know if I missed something. I'm open to suggestions. Thank you.

**Comment by aymanhab** [comment]
You can always reproduce the GUI/Edit behavior by passing the "bad" file to the constructor of ExternalLoads object then add it to a blank model and call initSystem on it, just make sure not all files are in the same folder you're running from.

**Comment by aseth1** [comment]
@ aymanhab I don't believe your analysis of the changes is correct. 

> Before 4.0 ExternalLoads was not a "Component" so we could create it from file, pass Model to the constructor and be done (edit ExternalForcess etc.)

ExternalLoads was being added to the Model before as it is now. The difference is only that we add the ExternalLoads to the Model as a single component (with subcomponents) instead of adding the individual ExternalForces within the ExternalLoads. This was all being done and remains to be done in `createExternalLoads` and this happens at precisely the same time. The change adheres to the interface of the Tool.

> In 4.0 ExternalLoads became a component so we were required to create it and add it to a model in some fashion and call initSystem on the model so we could use it. We could get away with that before the last change because invokeConnectToModel wasn't triggered then. This pathway was used by the GUI to load/edit/define ExternalLoads.

Again, this was the same as before. AbstractTool::createExternalLoads() behaves the same as it did in that it updates the Tool's `_model` to include `ExternalLoads` wired up for simulation/analysis. `initSystem()` does not come into play, except only to transform the point of forces application according supplied kinematics and even in this case it is on an copy of the model internal to the method.

> When refactoring ExternalLoads to be "Proper component", invokeConnectToModel was renamed and is now actually called through initSystem 

Again, being a Component and `initSystem` have no bearing on the issue.

> but that is done in context/working-directory different from the call in createExternalLoads (which changes working directory around the call) resulting in different behavior in GUI edit vs tool run.

Now this is the central and only issue AFAICT. The ExternalLoads::###connectToModel() are not called in the identical sequence. I see that the changing of the current working directory by `AbstractTool::createExternalLoads()` and yes that will affect the file loading. I think this can be easily remedied by invoking ExternalLoads::finalizeConnections in `AbstractTool::createExternalLoads()`. 

Thanks for point me to this small difference, that does seem to be the problem.

**Comment by aseth1** [comment]
The behavior in the Tool seems correct and the change to add an explicit `finalizeConnections` does not seem to change the behavior of the Tool. Therefore, the invocation of `Model::setup()` can be removed. The issue seems to be on fresh construction of `ExternalLoads`  in `EditExternalLoadsPanel.java`:

```java
        dLoads = new ExternalLoads(externalLoadsFilename, true);
        modelLocalCopy.addModelComponent(dLoads);
        dLoads.setName("Ex2");
        modelLocalCopy.initSystem();
```

and before it was:

```java
        dLoads = new ExternalLoads(modelLocalCopy, externalLoadsFilename);
        dLoads.setName("Ex2");
        modelLocalCopy.initSystem();
```

AFAICT we have only made the creation and addition of `ExternalLoad`s a two step process instead of a single step. I cannot see how these two code blocks are functionally different. One perplexing aspect, is when `###ConnectToModel()` was being called in the before case. It is now invoked via `initSystem()` and your comments above suggest it was not being invoked, but yet the data source was being loaded and that could only occur during `ExternalLoads::invokeConnectToModel()`. @aymanhab, was it being called directly in GUI prior to the set change? If so,  was the current working directory updated during that call?
I cannot find anything on the API other than in the specific Tools that change the current working directory.

**Comment by aseth1** [comment]
Fixed by #2332 Needs to be verified.
